// @mikro-orm/reflection/TsMorphMetadataProvider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TsMorphMetadataProvider = void 0;
const ts_morph_1 = require("ts-morph");
const core_1 = require("@mikro-orm/core");
class TsMorphMetadataProvider extends core_1.MetadataProvider {
    constructor() {
        super(...arguments);
        this.project = new ts_morph_1.Project({
            compilerOptions: {
                strictNullChecks: true,
                module: ts_morph_1.ModuleKind.Node16,
            },
        });
    }
    useCache() {
        return this.config.get('cache').enabled ?? true;
    }
    async loadEntityMetadata(meta, name) {
        if (!meta.path) {
            return;
        }
        await this.initProperties(meta);
    }
    async getExistingSourceFile(path, ext, validate = true) {
        if (!ext) {
            return await this.getExistingSourceFile(path, '.d.ts', false) || await this.getExistingSourceFile(path, '.ts');
        }
        const tsPath = path.match(/.*\/[^/]+$/)[0].replace(/\.js$/, ext);
        return (await this.getSourceFile(tsPath, validate));
    }
    async initProperties(meta) {
        // load types and column names
        for (const prop of Object.values(meta.properties)) {
            const type = this.extractType(prop);
            if (!type || this.config.get('discovery').alwaysAnalyseProperties) {
                await this.initPropertyType(meta, prop);
            }
            prop.type = type || prop.type;
        }
    }
    extractType(prop) {
        if (core_1.Utils.isString(prop.entity)) {
            return prop.entity;
        }
        if (prop.entity) {
            return core_1.Utils.className(prop.entity());
        }
        return prop.type;
    }
    async initPropertyType(meta, prop) {
        const { type, optional } = await this.readTypeFromSource(meta, prop);
        prop.type = type;
        if (optional) {
            prop.optional = true;
        }
        this.processWrapper(prop, 'IdentifiedReference');
        this.processWrapper(prop, 'Reference');
        this.processWrapper(prop, 'Ref');
        this.processWrapper(prop, 'Collection');
        if (prop.type.replace(/import\(.*\)\./g, '').match(/^(Dictionary|Record)<.*>$/)) {
            prop.type = 'json';
        }
    }
    async readTypeFromSource(meta, prop) {
        const source = await this.getExistingSourceFile(meta.path);
        const cls = source.getClass(meta.className);
        /* istanbul ignore next */
        if (!cls) {
            throw new core_1.MetadataError(`Source class for entity ${meta.className} not found. Verify you have 'compilerOptions.declaration' enabled in your 'tsconfig.json'. If you are using webpack, see https://bit.ly/35pPDNn`);
        }
        const properties = cls.getInstanceProperties();
        const property = properties.find(v => v.getName() === prop.name);
        if (!property) {
            return { type: prop.type, optional: prop.nullable };
        }
        const tsType = property.getType();
        const typeName = tsType.getText(property);
        if (prop.enum && tsType.isEnum()) {
            prop.items = tsType.getUnionTypes().map(t => t.getLiteralValueOrThrow());
        }
        if (tsType.isArray()) {
            prop.array = true;
            /* istanbul ignore else */
            if (tsType.getArrayElementType().isEnum()) {
                prop.items = tsType.getArrayElementType().getUnionTypes().map(t => t.getLiteralValueOrThrow());
            }
        }
        if (prop.array && prop.enum) {
            prop.enum = false;
        }
        let type = typeName;
        const union = type.split(' | ');
        const optional = property.hasQuestionToken?.() || union.includes('null') || union.includes('undefined');
        type = union.filter(t => !['null', 'undefined'].includes(t)).join(' | ');
        prop.array ?? (prop.array = type.endsWith('[]') || !!type.match(/Array<(.*)>/));
        type = type
            .replace(/Array<(.*)>/, '$1') // unwrap array
            .replace(/\[]$/, '') // remove array suffix
            .replace(/\((.*)\)/, '$1'); // unwrap union types
        // keep the array suffix in the type, it is needed in few places in discovery and comparator (`prop.array` is used only for enum arrays)
        if (prop.array && !type.includes(' | ') && prop.reference === core_1.ReferenceType.SCALAR) {
            type += '[]';
        }
        return { type, optional };
    }
    async getSourceFile(tsPath, validate) {
        if (!this.sources) {
            await this.initSourceFiles();
        }
        const source = this.sources.find(s => s.getFilePath().endsWith(core_1.Utils.stripRelativePath(tsPath)));
        if (!source && validate) {
            throw new core_1.MetadataError(`Source file '${tsPath}' not found. Check your 'entitiesTs' option and verify you have 'compilerOptions.declaration' enabled in your 'tsconfig.json'. If you are using webpack, see https://bit.ly/35pPDNn`);
        }
        return source;
    }
    processWrapper(prop, wrapper) {
        // type can be sometimes in form of:
        // `'({ object?: Entity | undefined; } & import("...").Reference<Entity>)'`
        // `{ object?: import("...").Entity | undefined; } & import("...").Reference<Entity>`
        // `{ node?: ({ id?: number | undefined; } & import("...").Reference<import("...").Entity>) | undefined; } & import("...").Reference<Entity>`
        // the regexp is looking for the `wrapper`, possible prefixed with `.` or wrapped in parens.
        const type = prop.type.replace(/import\(.*\)\./g, '');
        const m = type.match(new RegExp(`(?:^|[.( ])${wrapper}<(\\w+),?.*>(?:$|[) ])`));
        if (!m) {
            return;
        }
        prop.type = m[1];
        if (['Ref', 'Reference', 'IdentifiedReference'].includes(wrapper)) {
            prop.wrappedReference = true;
        }
    }
    async initSourceFiles() {
        // All entity files are first required during the discovery, before we reach here, so it is safe to get the parts from the global
        // metadata storage. We know the path thanks the the decorators being executed. In case we are running via ts-node, the extension
        // will be already `.ts`, so no change needed. `.js` files will get renamed to `.d.ts` files as they will be used as a source for
        // the ts-morph reflection.
        const paths = Object.values(core_1.MetadataStorage.getMetadata()).map(m => m.path.replace(/\.js$/, '.d.ts'));
        this.sources = this.project.addSourceFilesAtPaths(paths);
    }
}
exports.TsMorphMetadataProvider = TsMorphMetadataProvider;
